autoload/lsp/internal/semantic.vim

```diff
@@ -0,0 +1,354 @@
+
+function! s:init_highlight(server, buf) abort
+    let l:legend = lsp#internal#semantic#get_legend(a:server)
+    let l:highlight_groups = {
+        \ 'LspType': 'Type',
+        \ 'LspClass': 'Class',
+        \ 'LspEnum': 'Type',
+        \ 'LspInterface': 'TypeDef',
+        \ 'LspStruct': 'Type',
+        \ 'LspTypeParameter': 'Type',
+        \ 'LspParameter': 'Type',
+        \ 'LspVariable': 'Identifier',
+        \ 'LspProperty': 'TypeDef',
+        \ 'LspEnumMember': 'Constant',
+        \ 'LspEvents': 'Identifier',
+        \ 'LspFunction': 'Function',
+        \ 'LspMethod': 'Function',
+        \ 'LspKeyword': 'Keyword',
+        \ 'LspModifier': 'Type',
+        \ 'LspComment': 'Comment',
+        \ 'LspString': 'String',
+        \ 'LspNumber': 'Number',
+        \ 'LspRegexp': 'String',
+        \ 'LspOperator': 'Operator'
+    \ }
+
+    for l:token_name in l:legend['tokenTypes']
+        let l:hl_name = s:get_hl_name(l:token_name)
+        if !index(keys(l:highlight_groups), l:hl_name)
+            let l:highlight_groups[l:hl_name] = ''
+        endif
+    endfor
+
+    for [l:key, l:value] in items(l:highlight_groups)
+        if !hlexists(l:key)
+            if l:value !=# ''
+                exec 'highlight link' l:key l:value
+            else
+                exec 'highlight ' l:key 'gui=NONE cterm=NONE guifg=NONE ctermfg=NONE guibg=NONE ctermbg=NONE'
+            endif
+        endif
+    endfor
+
+    if s:use_vim_textprops
+        for l:token_idx in range(len(l:legend['tokenTypes']))
+            let l:token_name = l:legend['tokenTypes'][l:token_idx]
+            let l:hl = s:get_hl_name(l:token_name)
+            let l:textprop_name = s:get_textprop_name(a:server, l:token_idx)
+            silent! call prop_type_add(l:textprop_name, {'bufnr': a:buf, 'highlight': l:hl, 'combine': v:true})
+        endfor
+    endif
+endfunction
+
+function! s:clear_highlights(server, buf) abort
+    if s:use_vim_textprops
+        let l:legend = lsp#internal#semantic#get_legend(a:server)
+        for l:token_idx in range(len(l:legend['tokenTypes']))
+            let l:textprop_name = s:get_textprop_name(a:server, l:token_idx)
+            silent! call prop_remove({'bufnr': a:buf, 'type': l:textprop_name, 'all': v:true}, 1, line('$'))
+        endfor
+    elseif s:use_nvim_highlight
+        call nvim_buf_clear_namespace(a:buf, s:namespace_id, 0, line('$'))
+    endif
+endfunction
+
+function! s:add_highlight(server, buf, token) abort
+    let l:legend = lsp#internal#semantic#get_legend(a:server)
+    let l:startpos = lsp#utils#position#lsp_to_vim(a:buf, a:token['pos'])
+    let l:endpos = a:token['pos']
+    let l:endpos['character'] = l:endpos['character'] + a:token['length']
+    let l:endpos = lsp#utils#position#lsp_to_vim(a:buf, l:endpos)
+
+    try
+        if s:use_vim_textprops
+            let l:textprop_name = s:get_textprop_name(a:server, a:token['token_idx'])
+            call prop_add(l:startpos[0], l:startpos[1],
```

Comment:
we should use `prop_add_list` instead of `prop_add`?

https://github.com/prabirshrestha/vim-lsp/issues/1283